src/passes/OptimizeInstructions.cpp

```diff
@@ -1170,6 +1177,39 @@ struct OptimizeInstructions
   // Information about our locals
   std::vector<LocalInfo> localInfo;
 
+  // Check if two inputs to an instruction are equal, and can be removed. This
+  // is simpler than the general question of equality because we know they are
+  // inputs to the same instruction, and so they execute one after the other,
+  // with nothing else in the middle. So all we need to check is that they are
+  // structurally equal, and that they have no side effects (as if they did, we
+  // could not fold them together or otherwise remove any part of them).
+  bool equalAndRemovable(Expression* left, Expression* right) {
+    PassOptions passOptions = getPassOptions();
+    if (EffectAnalyzer(passOptions, getModule()->features, left)
+          .hasSideEffects() ||
+        EffectAnalyzer(passOptions, getModule()->features, right)
+          .hasSideEffects()) {
+      return false;
+    }
+    // Ignore extraneous things and compare.
+    left = Properties::getFallthrough(left, passOptions, getModule()->features);
+    right =
+      Properties::getFallthrough(right, passOptions, getModule()->features);
+    if (!ExpressionAnalyzer::equal(left, right)) {
+      return false;
+    }
+    // Reference equality also needs to check for allocation, which is *not* a
+    // side effect, but which results in different references:
+    // repeatedly calling (struct.new $foo)'s output will return different
+    // results (while i32.const etc. of course does not).
+    if (left->type.isRef()) {
+      if (FindAll<StructNew>(left).has() || FindAll<ArrayNew>(left).has()) {
+        return false;
+      }
+    }
```

Comment:
I think @tlively also pointed this out in the tests, but this doesn't seem to be able to rule out this case:
```wast
(ref.eq
  (block (result i32)
    (struct.new_default_with_rtt $struct
      (rtt.canon $struct)
    )
    (i32.const 0)
  )
  (block (result i32)
    (struct.new_default_with_rtt $struct
      (rtt.canon $struct)
    )
    (i32.const 0)
  )
)
```
This should be reduced to `(i32.const 1)` but it cannot because they contain allocations.